pgjdbc/src/main/java/org/postgresql/jdbc/PrimitiveArraySupport.java

```diff
@@ -0,0 +1,536 @@
+     */
+    @Override
+    public void appendArray(StringBuilder sb, char delim, boolean[] array) {
+      sb.append('{');
+      for (int i = 0; i < array.length; ++i) {
+        if (i > 0) {
+          sb.append(delim);
+        }
+        sb.append(array[i] ? '1' : '0');
+      }
+      sb.append('}');
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean supportBinaryRepresentation() {
+      return true;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @throws SQLFeatureNotSupportedException Because this feature is not supported.
+     */
+    @Override
+    public byte[] toBinaryRepresentation(boolean[] array) throws SQLFeatureNotSupportedException {
+      int length = 20 + (5 * array.length);
+      final byte[] bytes = new byte[length];
+
+      // 1 dimension
+      ByteConverter.int4(bytes, 0, 1);
+      // no null
+      ByteConverter.int4(bytes, 4, 0);
+      // oid
+      ByteConverter.int4(bytes, 8, Oid.BOOL);
+      // length
+      ByteConverter.int4(bytes, 12, array.length);
+
+      int idx = 20;
+      for (int i = 0; i < array.length; ++i) {
+        bytes[idx + 3] = 1;
+        ByteConverter.bool(bytes, idx + 4, array[i]);
+        idx += 5;
+      }
+
+      return bytes;
+    }
+
+  };
+
+  private static final ArrayToString<String[]> STRING_ARRAY_TOSTRING = new ArrayToString<String[]>() {
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int getDefaultArrayTypeOid() {
+      return Oid.VARCHAR_ARRAY;
+    }
+
+    @Override
+    public String toArrayString(char delim, String[] array) {
+      final StringBuilder sb = new StringBuilder(Math.max(64, array.length * 8));
+      appendArray(sb, delim, array);
+      return sb.toString();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void appendArray(StringBuilder sb, char delim, String[] array) {
+      sb.append('{');
+      for (int i = 0; i < array.length; ++i) {
+        if (i > 0) {
+          sb.append(delim);
+        }
+        if (array[i] == null) {
+          sb.append('N');
+          sb.append('U');
+          sb.append('L');
+          sb.append('L');
+        } else {
+          PgArray.escapeArrayElement(sb, array[i]);
+        }
+      }
+      sb.append('}');
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean supportBinaryRepresentation() {
+      return false;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @throws SQLFeatureNotSupportedException Because this feature is not supported.
+     */
+    @Override
+    public byte[] toBinaryRepresentation(String[] array) throws SQLFeatureNotSupportedException {
+      throw new SQLFeatureNotSupportedException();
+    }
+
+  };
+
+  private static final Map<Class, ArrayToString> ARRAY_CLASS_TOSTRING = new HashMap<Class, ArrayToString>(9);
+
+  static {
+    ARRAY_CLASS_TOSTRING.put(long[].class, LONG_ARRAY_TOSTRING);
+    ARRAY_CLASS_TOSTRING.put(int[].class, INT_ARRAY_TOSTRING);
```

Comment:
No, multi-dimensional arrays continue to always use string representation. However, the string generation is optimized to operate with primitive ints, rather than using reflections, auto-boxing and string escaping.